File: skillet/src/main/java/dev/kkorolyov/pancake/skillet/model/Workspace.java
package dev.kkorolyov.pancake.skillet.model;

import dev.kkorolyov.pancake.skillet.model.factory.ComponentFactory;

import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A self-contained context representing a single Skillet project.
 */
public class Workspace extends Model<Workspace> {
	private final ComponentFactory componentFactory = new ComponentFactory();

	private final Collection<GenericEntity> entities = new LinkedHashSet<>();
	private GenericEntity activeEntity;

	/** @return workspace component factory */
	public ComponentFactory getComponentFactory() {
		return componentFactory;
	}

	/**
	 * Adds all entities in a workspace to this workspace.
	 * @param workspace workspace with entities to add
	 * @return {@code this}
	 */
	public Workspace addWorkspace(Workspace workspace) {
		workspace.streamEntities()
				.forEach(this::addEntity);
		return this;
	}

	/**
	 * Adds a new entity to the workspace.
	 * @return new entity
	 */
	public GenericEntity newEntity() {
		GenericEntity entity = new GenericEntity();
		addEntity(entity);

		return entity;
	}

	/** @return {@code true} if this workspace contains {@code entity} */
	public boolean containsEntity(GenericEntity entity) {
		return entities.contains(entity);
	}

	/**
	 * Adds an entity and sets it as the active entity.
	 * Adds all new components in {@code entity} to the component factory.
	 * @param entity entity to add
	 * @return {@code this}
	 */
	public Workspace addEntity(GenericEntity entity) {
		entities.add(entity);
		setActiveEntity(entity);

		componentFactory.add(entity.getComponents(), false);

		changed(WorkspaceChangeEvent.ADD);

		return this;
	}
	/**
	 * Attempts to remove an entity from the workspace.
	 * @param entity entity to remove
	 * @return {@code true} if this workspace contained {@code entity} and it was removed
	 */
	public boolean removeEntity(GenericEntity entity) {
		boolean result = entities.remove(entity);

		if (result) changed(WorkspaceChangeEvent.REMOVE);

		return result;
	}

	/** @return all entities in this workspace */
	public Iterable<GenericEntity> getEntities() {
		return entities;
	}
	/** @return stream over all entities */
	public Stream<GenericEntity> streamEntities() {
		return entities.stream();
	}

	/** all components of all entities */
	public Iterable<GenericComponent> getComponents() {
		return streamComponents().collect(Collectors.toList());
	}
	/** @return stream over all components of all entities */
	public Stream<GenericComponent> streamComponents() {
		return streamEntities().flatMap(GenericEntity::streamComponents);
	}

	/** @return last entity set as active */
	public Optional<GenericEntity> getActiveEntity() {
		return Optional.ofNullable(activeEntity);
	}
	/**
	 * @param entity entity to set as active
	 * @throws IllegalArgumentException if {@code entity} is not in this workspace
	 */
	public void setActiveEntity(GenericEntity entity) {
		if (Objects.equals(activeEntity, entity)) return;

		if (entity != null && !containsEntity(entity)) throw new IllegalArgumentException("Active entity must be in the workspace");

		this.activeEntity = entity;
		changed(WorkspaceChangeEvent.ACTIVE);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) return true;
		if (obj == null || !getClass().isAssignableFrom(obj.getClass())) return false;

		Workspace o = (Workspace) obj;
		return Objects.equals(componentFactory, o.componentFactory)
				&& Objects.equals(entities, o.entities)
				&& Objects.equals(activeEntity, o.activeEntity);
	}
	@Override
	public int hashCode() {
		return Objects.hash(componentFactory,
				entities,
				activeEntity);
	}
	/**
	 * A change to a workspace.
	 */
	public enum WorkspaceChangeEvent implements ModelChangeEvent {
		ACTIVE,
		ADD,
		REMOVE
	}
}

File: skillet/src/main/kotlin/dev/kkorolyov/pancake/skillet/model/Workspace.kt
package dev.kkorolyov.pancake.skillet.model

import dev.kkorolyov.pancake.skillet.model.factory.ComponentFactory
import java.util.Collections
import java.util.IdentityHashMap
import java.util.Objects

/**
 * A self-contained context representing a single Skillet project.
 */
class Workspace : Model<Workspace>() {
	/** Attached component factory */
	val componentFactory: ComponentFactory = ComponentFactory()

	private val _entities: MutableCollection<GenericEntity> = Collections.newSetFromMap(IdentityHashMap())
	/** All entities */
	val entities: Iterable<GenericEntity> get() = _entities
	/** Components of all entities */
	val components: Iterable<GenericComponent> get() = entities.flatMap(GenericEntity::components)

	var activeEntity: GenericEntity? = null
		/**
		 * @param entity entity to set as active
		 * @throws IllegalArgumentException if `entity` is not in this workspace
		 */
		set(entity) {
			if (field === entity) return

			if (entity !== null && entity !in this) throw IllegalArgumentException("Active entity ($entity) must be in the workspace")

			field = entity
			changed(WorkspaceChangeEvent.ACTIVE)
		}

	/**
	 * Adds all entities from a workspace.
	 * @param workspace workspace with entities to add
	 * @return `this`
	 */
	fun addWorkspace(workspace: Workspace): Workspace {
		workspace._entities.forEach { add(it) }
		return this
	}

	/**
	 * Creates and adds a new entity.
	 * @return new entity
	 */
	fun newEntity(): GenericEntity {
		val entity = GenericEntity()
		add(entity)

		return entity
	}

	/** @return `true` if this workspace contains `entity` */
	operator fun contains(entity: GenericEntity): Boolean = _entities.contains(entity)

	/** @see [add] */
	operator fun plusAssign(entity: GenericEntity) {
		add(entity)
	}
	/**
	 * Adds an entity and sets it as the active entity.
	 * Adds all new components in `entity` to the component factory.
	 * @param entity entity to add
	 * @return `this`
	 */
	fun add(entity: GenericEntity): Workspace {
		_entities.add(entity)
		activeEntity = entity

		componentFactory.add(entity.components, false)

		changed(WorkspaceChangeEvent.ADD)

		return this
	}
	/**
	 * Attempts to remove an entity from this workspace.
	 * @param entity entity to remove
	 * @return `true` if this workspace contained `entity` and it was removed
	 */
	fun removeEntity(entity: GenericEntity): Boolean {
		val result = _entities.remove(entity)

		if (result) changed(WorkspaceChangeEvent.REMOVE)

		return result
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other === null || !this::class.isInstance(other)) return false

		other as Workspace
		return componentFactory == other.componentFactory
				&& _entities == other._entities
				&& activeEntity == other.activeEntity
	}
	override fun hashCode(): Int = Objects.hash(componentFactory,
			_entities,
			activeEntity)

	/**
	 * A change to a workspace.
	 */
	enum class WorkspaceChangeEvent : ModelChangeEvent {
		ACTIVE,
		ADD,
		REMOVE
	}
}
